### PR TITLE
Introduce country-specific address validation

### PR DESCRIPTION
**Pull request by raphaelm**

### CODE REVIEW

File: src/pretix/base/addressvalidation.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+from pretix.base.settings import COUNTRIES_WITH_STATE_IN_ADDRESS
+
+_validator_classes = defaultdict(list)
+
+COUNTRIES_WITH_STREET_ZIPCODE_AND_CITY_REQUIRED = {
+    # We don't presume this for countries we don't have knowledge about, there are countries in the
+    # world e.g. without zipcodes
+    'AR', 'AT', 'AU', 'BE', 'BR', 'CA', 'CH', 'CN', 'CU', 'CZ', 'DE', 'DK', 'EE', 'ES', 'FI', 'FR',
+    'GB', 'GR', 'HR', 'ID', 'IE', 'IL', 'IN', 'IR', 'IS', 'IT', 'JP', 'LT', 'LV', 'MA', 'MT', 'MX',
+    'NL', 'NO', 'NZ', 'PK', 'PL', 'PT', 'RO', 'RU', 'SE', 'SI', 'SG', 'SK', 'TR', 'UA', 'US', 'ZA',
```
Comment (wiffbi) [inline src/pretix/base/addressvalidation.py:84]: Easier to compare with imports when used in same/alphabetical order.
```suggestion
    'NL', 'NO', 'NZ', 'PK', 'PL', 'PT', 'RO', 'RU', 'SE', 'SG', 'SI', 'SK', 'TR', 'UA', 'US', 'ZA',
```

File: src/pretix/base/addressvalidation.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    :return: The dictionary, possibly with changes
+    """
+    if not address.get('street') and not address.get('zipcode') and not address.get('city'):
+        # Consider the actual address part to be empty, no further validation necessary, if the
+        # address should be required, it's the callers job to validate that at least one of these
+        # fields is filled
+        return address
+
+    if not address.get('country'):
+        raise ValidationError({'state': [_('This field is required.')]})
```
Comment (wiffbi) [inline src/pretix/base/addressvalidation.py:101]: A little typo:

```suggestion
        raise ValidationError({'country': [_('This field is required.')]})
```

**Review by wiffbi — CHANGES_REQUESTED** [review]
Sorry, thought it would be just comments, but could not get it to work due to `validate_required_fields` not being defined on `BaseValidator`. Do I miss something?

File: src/pretix/base/addressvalidation.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    if str(address['country']) in COUNTRIES_WITH_STREET_ZIPCODE_AND_CITY_REQUIRED:
+        for f in ('street', 'zipcode', 'city'):
+            if not address.get(f):
+                raise ValidationError({f: [_('This field is required.')]})
+
+    for klass in _validator_classes[str(address['country'])]:
+        validator = klass()
+        validator.validate_required_fields(address)
```
[the remark this answers is not in the document]
Comment (raphaelm) [reply src/pretix/base/addressvalidation.py:113]: Oops, sorry, that got removed during refactoring. Still need to create tests for this :)

**Review by wiffbi — APPROVED** [review]

**Pull request merged**